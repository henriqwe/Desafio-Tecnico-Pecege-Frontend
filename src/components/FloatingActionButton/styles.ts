import styled from 'styled-components'

const variantOptions = {
  left: {
    left: '1rem'
  },
  right: {
    right: '1rem'
  }
}

export const FloatingActionButtonContainer = styled.div<{
  position: 'left' | 'right'
}>`
  position: absolute;
  display: flex;
  width: 5rem;
  border-radius: 100%;
  align-items: center;
  justify-content: center;
  height: 5rem;
  bottom: 1rem;
  background-color: #ffaf49;
  transition: all 100ms ease-in-out;
  box-shadow: 0 2px 5px rgba(0, 0, 0, 0.25);
  &:hover {
    background-color: #e5993b;
    cursor: pointer;
    width: 5.5rem;
    height: 5.5rem;
  }

  @media (min-width: 768px) {
    height: 3rem;
    width: 3rem;
    &:hover {
      background-color: #e5993b;
      cursor: pointer;
      width: 3.5rem;
      height: 3.5rem;
    }
  }

  ${({ position }) => variantOptions[position]}
`

export const WrapperIcon = styled.div`
  width: 3rem;
  height: 3rem;
  display: flex;
  justify-content: center;
  align-items: center;
  transition: all 0.5s;
`

export const IconWrapper = styled.div`
  width: 3rem;
  height: 3rem;
  color: ${({ theme }) => theme.colors.white};
  @media (min-width: 768px) {
    width: 2rem;
    height: 2rem;
  }
`
