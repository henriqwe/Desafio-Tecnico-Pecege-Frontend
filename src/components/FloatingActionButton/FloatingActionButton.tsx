import * as S from './styles'
import { MapPin } from '&components/Icons/MapPin'
import { useLocation, useNavigate } from 'react-router-dom'
import { routes } from '&utils/routes'
import { ChevronLeftIcon } from '@heroicons/react/24/outline'
import { motion } from 'framer-motion'

export function FloatingActionButton() {
  const navigate = useNavigate()
  const { pathname } = useLocation()
  const { icon, position, show, title, pathTo } = handlepath(pathname)

  if (!show || !pathTo || !position) {
    return <></>
  }
  return (
    <motion.div
      initial={{ opacity: 0 }}
      animate={{ opacity: 1 }}
      exit={{ opacity: 0 }}
      style={{ width: '100%', height: '100%' }}
      transition={{
        type: 'spring',
        stiffness: 260,
        damping: 20
      }}
    >
      <S.FloatingActionButtonContainer
        title={title}
        onClick={() => navigate(pathTo)}
        position={position}
        data-testid={'floatingactionbutton-container'}
      >
        <S.WrapperIcon>{icon}</S.WrapperIcon>
      </S.FloatingActionButtonContainer>
    </motion.div>
  )
}

function handlepath(pathname: string) {
  let show = true
  const options: {
    [x: string]: {
      route: string
      position: 'right' | 'left'
      icon: JSX.Element
      title: string
      pathTo: string
    }
  } = {
    home: {
      route: routes.home.path,
      position: 'right',
      icon: (
        <S.IconWrapper>
          <MapPin />
        </S.IconWrapper>
      ),
      title: 'Ir para mapa',
      pathTo: routes.map.path
    },
    map: {
      route: routes.map.path,
      position: 'left',
      icon: (
        <S.IconWrapper>
          <ChevronLeftIcon />
        </S.IconWrapper>
      ),
      title: 'Ir para mapa',
      pathTo: routes.home.path
    }
  }

  const path = Object.values(options).find((path) => path.route === pathname)

  if (!path) {
    show = false
    return { show }
  }
  const { icon, position, title, pathTo } = path

  return { show, icon, position, title, pathTo }
}
