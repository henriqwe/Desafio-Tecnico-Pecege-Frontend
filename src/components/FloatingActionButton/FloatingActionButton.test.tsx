import { fireEvent, render, waitFor } from '@testing-library/react'
import '@testing-library/jest-dom'
import { FloatingActionButton } from './FloatingActionButton'
import { TestProviders } from '&utils/tests/TestProviders'

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: () => ({
    pathname: '/'
  })
}))
describe('FloatingActionButton', () => {
  it('should render correctly snapshot', async () => {
    const { container, findByTestId } = render(
      <TestProviders>
        <FloatingActionButton />
      </TestProviders>
    )

    const floatingActionButtonContainer = await findByTestId(
      'floatingactionbutton-container'
    )
    await waitFor(async () => fireEvent.click(floatingActionButtonContainer))

    expect(container).toMatchSnapshot()
  })
})
