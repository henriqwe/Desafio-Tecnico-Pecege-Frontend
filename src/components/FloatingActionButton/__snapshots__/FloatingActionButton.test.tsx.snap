// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatingActionButton should render correctly snapshot 1`] = `
<div>
  <div
    class="sc-bcPKhP gXbqDV"
    data-testid="floatingactionbutton-container"
    position="right"
    title="Ir para mapa"
  >
    <div
      class="sc-grXZZQ dzrPzn"
    >
      <div
        class="sc-dkjKgF dEavTc"
      >
        <svg
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            color="white"
            d="M12 18.5L9 17l-6 3V7l6-3l6 3l6-3v7M9 4v13m6-10v5m6.121 8.121a3 3 0 1 0-4.242 0c.418.419 1.125 1.045 2.121 1.879c1.051-.89 1.759-1.516 2.121-1.879zM19 18v.01"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
